### PR TITLE
Refactor + Hooks to support Visualizing the tree nodes.

### PR DESCRIPTION
**Pull request by stefanpenner**
* formalize Node into a first class entity
* extract and name large callbacks in then chains
* remove semi-colon
* rename self -> builder so it is clear when reading
* RSVP.resolve -> Promise.resolve
* upgrade RSVP
* add scripts/graph.js which converts the broccoli-tree.json into dot for graphviz to consume

all this provide the hooks to produce the following visualization: (https://github.com/stefanpenner/broccoli-viz will provide the rest)

1. name of named trees
2. id un-named of tree
3. self time of each tree

graph

### CODE REVIEW

**Comment by krisselden** [comment]
I think having some sort of inspect() would be incredibly useful, it isn't always easy to suss out a tree.  Can we please move this forward somehow?

**Comment by stefanpenner** [comment]
rebased

**Comment by joliss** [comment]
I can't pull this in right now - between the upcoming Builder rewrite and the multiple releases there's too many spinning plates. :(

When I write the new Builder I'll be sure to go back to this PR and copy the JSON API you implemented.

Sorry for making all your awesome visualization work block on my end!

**Comment by stefanpenner** [comment]
Just to point this out, this has been blocked for 4-5 months now. Thats 4-5 months, where we had poor internal debugging experiences, and no reasonable way to explore further visualization work.

What is wrong with merging it, and then upgrading as the builder changes? This lag is killing my motivation to solve issues in this project. I am just more and more inclined to just fork and leave an eternal pull request for some theoretical future merge. That is that last thing I want to do, but I'm hurting here.

**Comment by joliss** [comment]
As for merging *right now*: I can, but if I merge it now it'd be released in v0.15.1 (unless I want to start branching stuff), which isn't usable for an ember-cli release yet - you could only use it on master. Also the API would almost certainly change within the next few weeks as the builder gets redone. Do you still want me to merge it?

As for blocking 4-5 months: Yup, that's broken - I feel your pain. I want to fix that.

**Comment by joliss** [comment]
> I want to fix that.

As in, I think/hope it'll be better now that I spend time on Broccoli in regular weekly intervals.

**Comment by stefanpenner** [comment]
> I want to fix that.
> As in, I think/hope it'll be better now that I spend time on Broccoli in regular weekly intervals.

good to here.

I am maintaining https://github.com/stefanpenner/broccoli/tree/ember-cli for now, and am working to cleanup some of the active PR's to broccolijs as i refactor them in the new setup. We can consider that a bleeding edge experimental branch, that lets us iterate faster. Hopes are that it can be a proving ground.

Currently it is tracking 0.13.x, until the latest refactorings enter the world of broccoli.

File: package.json [hunk 1]
```diff
@@ -19,18 +19,20 @@
...
     "handlebars": "^2.0.0",
+    "jshint": "^2.5.10",
     "mime": "^1.2.11",
-    "promise-map-series": "^0.2.1",
     "quick-temp": "^0.1.2",
     "rimraf": "^2.2.8",
-    "rsvp": "^3.0.6",
+    "promise-map-series": "^0.2.1",
+    "rsvp": "^3.0.14",
+    "tap": "^0.4.13",
```
Comment (joliss) [inline package.json]: Hm, some packages seem to have been moved from devDependencies to dependencies. Are these upgrades actually necessary?

Comment (stefanpenner) [inline package.json]: tap isn't, let me fix

**Comment by stefanpenner** [comment]
@joliss rebased, now that ember-cli is running the latest (or near latest) that should address your last concern

**Comment by joliss** [comment]
Thanks, merged and released in 0.16.1!

**Comment by stefanpenner** [comment]
got broccoli-viz working again \w latest broccoli. 

https://github.com/stefanpenner/broccoli-viz

**Pull request merged**